src/backend/distributed/commands/multi_copy.c

```diff
@@ -204,20 +218,117 @@ CitusCopyFrom(CopyStmt *copyStatement, char *completionTag)
 		}
 	}
 
-	partitionMethod = PartitionMethod(tableId);
-	if (partitionMethod == DISTRIBUTE_BY_HASH || partitionMethod == DISTRIBUTE_BY_RANGE)
+	isCopyFromWorker = IsCopyFromWorker(copyStatement);
+	if (isCopyFromWorker)
+	{
+		CopyFromWorkerNode(copyStatement, completionTag);
+	}
+	else
+	{
+		char partitionMethod = PartitionMethod(tableId);
+		if (partitionMethod == DISTRIBUTE_BY_HASH || partitionMethod ==
+			DISTRIBUTE_BY_RANGE)
+		{
+			CopyToExistingShards(copyStatement, completionTag);
+		}
+		else if (partitionMethod == DISTRIBUTE_BY_APPEND)
+		{
+			CopyToNewShards(copyStatement, completionTag);
+		}
+		else
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("unsupported partition method")));
+		}
+	}
+}
+
+
+/*
+ * IsCopyFromWorker checks if the given copy statement has the master host option.
+ */
+bool
+IsCopyFromWorker(CopyStmt *copyStatement)
+{
+	ListCell *optionCell = NULL;
+	foreach(optionCell, copyStatement->options)
 	{
-		CopyToExistingShards(copyStatement, completionTag);
+		DefElem *defel = (DefElem *) lfirst(optionCell);
+		if (strncmp(defel->defname, "master_host", NAMEDATALEN) == 0)
+		{
+			return true;
+		}
 	}
-	else if (partitionMethod == DISTRIBUTE_BY_APPEND)
+
+	return false;
+}
+
+
+/*
+ * CopyFromWorkerNode implements the COPY table_name FROM ... from worker nodes
+ * for append-partitioned tables.
+ */
+static void
+CopyFromWorkerNode(CopyStmt *copyStatement, char *completionTag)
+{
+	NodeAddress *masterNodeAddress = MasterNodeAddress(copyStatement);
+	char *nodeName = masterNodeAddress->nodeName;
+	int32 nodePort = masterNodeAddress->nodePort;
+	char *nodeUser = CurrentUserName();
+
+	masterConnection = ConnectToNode(nodeName, nodePort, nodeUser);
+
+	PG_TRY();
 	{
+		PGresult *queryResult = NULL;
+		char *relationName = copyStatement->relation->relname;
+
+		char partitionMethod = MasterPartitionMethod(relationName);
+		if (partitionMethod != DISTRIBUTE_BY_APPEND)
+		{
+			ereport(ERROR, (errmsg("copy from worker nodes are only supported "
+								   "for append-partitioned tables")));
+		}
+
+		/* run all metadata commands in a transaction */
+		queryResult = PQexec(masterConnection, "BEGIN");
+		if (PQresultStatus(queryResult) != PGRES_COMMAND_OK)
+		{
+			ereport(ERROR, (errmsg("failed to start updating the metadata on the "
```

Comment:
this error message feels a little weird, but I suppose it will almost never happen anyway :)